Experimental Persistent KV store

mriise opened the pull request:
This adds a generic persistent kv interface for server side values. The intent is this will eventually be exposed to clients.

The current backing database is SQLite.

DATABASE STRUCTURE UNSTABLE! You may lose data until it is considered stable.

- There is no networking (clients cant use it in any way atm)
- There is no way to configure limits for buckets/keys or database settings
- Performance is not measured
- Tests are AI generated (take with a grain of salt)

~~@dooly123 DO NOT MERGE DRAFT!~~

mriise committed:
feat: Experimental Persistient KV store

This adds a generic persistent kv interface for server side values. The intent is this will eventually be exposed to clients.

The current backing database is SQLite.

DATABASE STRUCTURE UNSTABLE! You may lose data until it is considered stable.

- There is no networking (clients cant use it in any way atm)
- There is no way to configure limits for buckets/keys or database settings
- Performance is not measured
- Tests are AI generated (take with a grain of salt)

TheButlah reviewed (commented):
Seems very useful :3
how difficult would it be to get the KV store to work on either client or server rather than only supporting server KVs? One could imagine keys being labeled as persisted either locally (and are client-authoritative) or on the server (and are server-authoritative).

mriise commented:
we would need: 
a) strong embedded db (sqlite preferably)
b) stable format for that db

in reality you dont need a sync mechanism just to _have_ a client kv store (we can do it independently), but for what you are asking we would.

In many cases though i would think that you just want to keep things in-memory, then persist on exit of the app or instance. That could just as easily be done with file serialization.
this last point i would defer to @dooly123 on however.

TheButlah commented:
> you just want to keep things in-memory, then persist on exit of the app or instance

That same line of reasoning could be applied to indicate that we should not use sqlite on the server, and instead just persist things on exit as json. But in reality, I think its beneficial to have something flushed to disk when the function exits, rather than when the program exists, and sqlite is a known and reliable way to transactionally do that.

The same rationale that justifies its use on the server, justifies its use on the client (at least, imo)

mriise committed:
feat: kv store performance & testing
- a lot of ai generated tests.
- remove double try/catch
- sanitize prefix, make case sensitive, and sort results (ordinal).
- change how size usage is calculated.
- auto increment kv timestamp & version.
- use defer tx for read heavy tx.
- use sync transactions, microsoft reccomends against using the async apis for sqlite.
- Use a single, shared connection with a mutex. Despite having "connection pooling" for sqlite, making new connections significantly harm performance. Sqlite doesnt really allow for more than one writer anyway.

mriise committed:
remove caching and pooling from the connection string constructor, use pragma only

mriise committed:
return error on too long prefix for list keys, comment out management interface for now

mriise committed:
chore: cleanup .csproj

mriise committed:
refactor: comment out more unused things, split stat tracking to its own fn

mriise committed:
refactor: rename folders, namespaces, and split up KvStore into seperate files

mriise committed:
chore: add a few more tests

mriise reviewed (commented):
Tests are a bit verbose, but fine enough. Some of them reference the static class, so that will need updating when another backend is added (and a management interface is settled)

dooly123 committed:
fix for source files exporter including server only persistent KV store

Kitlith I'm sorry
